### PR TITLE
SWS-71 Fetch some of the prometheus metrics

### PR DESCRIPTION
**Pull request by jotak**
Still TODO:

- [x] Write tests with testify
- [x] Rebase & update when SWS-62 (rest endpoints) is merged

### CODE REVIEW

File: prometheus/client.go [hunk 1]
```diff
@@ -14,14 +15,33 @@ import (
...
 
-const istio_request_query = "istio_request_count{destination_service=~\"%s.%s.*\"}"
-
 // Client for Prometheus API.
 // It hides the way we query Prometheus offering a layer with a high level defined API.
 type Client struct {
 	p8s api.Client
 }
 
+// Metric is a base interface for either MetricValue (single scalar, gauge or counter) or MetricHistogram
```
Comment (jotak) [inline prometheus/client.go]: Is it the place for model types or should we have a dedicated package?

Comment (lucasponce) [reply prometheus/client.go]: I put in the client, the wrapper for the connection and basic ones, if it grows as it happens in kubernetes/istio I separated the files.

File: prometheus/client.go [hunk 2]
```diff
@@ -42,7 +62,8 @@ func NewClient() (*Client, error) {
 // Destination service is not included in the map as it is passed as argument.
 // It returns an error on any problem.
 func (in *Client) GetSourceServices(namespace string, servicename string) (map[string]string, error) {
-	query := fmt.Sprintf(istio_request_query, servicename, namespace)
+	query := fmt.Sprintf("istio_request_count{destination_service=\"%s.%s.%s\"}",
+		servicename, namespace, config.Get().Istio_Identity_Domain)
```
Comment (lucasponce) [inline prometheus/client.go]: Ok, if you introduce the Istio_Identity_Domain I guess that then it makes sense to filter it in the method, a couple of lines  below there is a 
```
			if i := strings.Index(sourceService, ".svc"); i > 0 {
				sourceService = sourceService[:i]
			}
```
that should be updated with the specific variable.

File: config/config.go [hunk 3]
```diff
@@ -44,6 +45,7 @@ type Config struct {
 	Identity               security.Identity ",omitempty"
 	Server                 Server            ",omitempty"
 	Prometheus_Service_Url string            ",omitempty"
+	Istio_Identity_Domain  string            ",omitempty"
 }
```
Comment (jshaughn) [inline config/config.go]: trivial: I'm no expert but I'm fairly sure the Go convention for naming encourages MixedCase as opposed to all caps or underscores.  Perhaps we should be using names like:

EnvServerAddress
PrometheusServiceUrl

I'm not sure, just looking at some other examples and https://golang.org/doc/effective_go.html#names

File: prometheus/client.go [hunk 4]
```diff
@@ -66,7 +87,147 @@ func (in *Client) GetSourceServices(namespace string, servicename string) (map[s
...
+	if err != nil {
+		return nil, err
+	}
+	return &MetricValue{
+		Query: query,
+		Value: val}, nil
+}
+
+func fetchHistogram(api v1.API, metricName string, labels string, duration string, now time.Time) (*MetricHistogram, error) {
+	// Note: we may want to make returned stats configurable in the future
```
Comment (jshaughn) [inline prometheus/client.go]: +1, it seems Prom handles each quantile as a separate query so there is some re-work for every requested quantile.  Also, we could, I think, blast of the requested quantile queries in parallel as goroutines.

File: config/config.go (hunk 3 above)
Comment (lucasponce) [reply config/config.go]: This is due the yaml parser, to read prometheus-service-url.

Comment (jmazzitelli) [reply config/config.go]: This is why there is a all-caps comment on this object about "USED FOR YAML"

Because the Yaml parser cannot be told in the ",omitempty" section what the yaml property is to be called, it is forced to determine this from the variable name. And how it does this is simply make the variable name all lowercase with no demarcation characters like - or _.

So if we do not name them with underscores, the words in the property names will all run together --- if we did it with your suggestion, for example, the yaml file would have properties called "istioidentitydomain" and "prometheusserviceurl". Difficult to read. So we use underscores between words so when you write the yaml file the properties are "istio_identity_domain" and "prometheus_service_url" - much easier to read and understand.

So in short the variable names in the structs that are used for YAML processing have _ between words. Look for the comment at the top of structs - if you see USED FOR YAML then we will use this naming convention. Otherwise, we do use camel case.

Comment (jshaughn) [reply config/config.go]: OK, thanks. I think actually we talked about this off-pr but I forgot.  The PR diff didn't show that comment so I didn't see it (although the comment doesn't mention that this unconventional format is needed).  Anyway, no problem, carry on.

Comment (simonpasquier) [reply config/config.go]: Unless I'm missed something obvious, you could use struct tags to stick with the Golang conventions:

```
type Config struct {
	Identity             security.Identity ",omitempty"
	Server               Server            ",omitempty"
	PrometheusServiceUrl string            "prometheus_service_url,omitempty"
	IstioIdentityDomain  string            "istio_identity_domain,omitempty"
}
```

It's widely used in the Prometheus code base, see https://github.com/prometheus/prometheus/blob/master/config/config.go#L267-L279 for instance.

Comment (jmazzitelli) [reply config/config.go]: AFAIK, this was not supported in the version of yaml parser at the time I originally started using this yaml parser. If this is now a supported feature of the yaml parser, then yes, this is what we should do.

Comment (jmazzitelli) [reply config/config.go]: https://godoc.org/gopkg.in/yaml.v2

Yes, looks supported. Feel free to refactor our config structures with this.

Comment (josejulio) [reply config/config.go]: If no one is doing it already, I can refactor the config structures.

Comment (jotak) [reply config/config.go]: @jshaughn vscode shows me warning about it, but I preferred to stick with the rules adopted with the other settings.

In a similar way it is also showing warning for all the const defined above (like ENV_IDENTITY_CERT_FILE) : "don't use ALL_CAPS in Go names; use CamelCase"

Comment (josejulio) [reply config/config.go]: I sent a PR about the config structures: https://github.com/swift-sunshine/swscore/pull/33

File: prometheus/client_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/swift-sunshine/swscore/config"
+)
+
+func setupConfig() {
+	conf := config.NewConfig()
+	conf.PrometheusServiceUrl = "http://prometheus-istio-system.127.0.0.1.nip.io"
+	config.Set(conf)
+}
+
+// For now it's just a runnable manual test rather than an itest.
```
Comment (lucasponce) [inline prometheus/client_test.go]: I also thought on these kind of manual tests.
I wonder if there is a way to uncomment them and just enable them with some flag on the config or prompt i.e. 'Manual/Whatever'.
In that way perhaps it would save time if you want to use these tests.

I don't know which is the best practice in golang, in maven we used to have these kind of variables for those manual tests.

File: prometheus/client.go [hunk 6]
```diff
@@ -66,7 +65,28 @@ func (in *Client) GetSourceServices(namespace string, servicename string) (map[s
 	return routes, nil
 }
 
+// GetServiceMetrics returns a map of metrics (indexed by description) related to the provided service identified by its namespace and service name.
+// Returned map includes istio metrics and envoy health.
+// It returns an error on any problem.
+func (in *Client) GetServiceMetrics(namespace string, servicename string, duration string) (map[string]Metric, error) {
+	metricChan, errChan := make(chan namedMetric), make(chan error)
```
Comment (lucasponce) [inline prometheus/client.go]: Missing close of the channels before to leave the function ?
(not sure if perhaps it's not necessary, but just come to my attention).

Comment (burmanm) [reply prometheus/client.go]: Closing channels makes code easier to read often though, since one can use for item := range metricChan { .. process items .. } just like it would have been a list. No need to return the amount of items or do the for loop against item count like later here. This is just a generic note, in this case two channels are read to it might make sense, although one could just loop over the errChan first and then quit without reading the other channel.

File: prometheus/client.go [hunk 7]
```diff
@@ -66,7 +65,28 @@ func (in *Client) GetSourceServices(namespace string, servicename string) (map[s
...
+	var err error
+	for i := 0; i < nbCalls; i++ {
+		select {
+		case namedMetric := <-metricChan:
+			metrics[namedMetric.name] = namedMetric.metric
+		case e := <-errChan:
+			err = e
+		}
+	}
+	return metrics, err
```
Comment (burmanm) [inline prometheus/client.go]: Does it make sense to return items and errors to the caller? Or just the error? Is the intention to keep processing the items even if an error happens (and ignore missing values) ?

File: prometheus/client_test.go (hunk 5 above)
Comment (burmanm) [reply prometheus/client_test.go]: I think in these tests we should just mock the prometheus server. A simple testHttpServer would do the job and return some metrics (which can be then asserted against).

File: prometheus/client.go (hunk 7 above)
Comment (jotak) [reply prometheus/client.go]: Well I think it's nicer if we can provide the information we got even if some queries failed (like because of an unstable network, for instance)

**Comment by lucasponce** [comment]
What would need this PR to progress ?
4 days and 6 people reviewing it sounds a lot.

**Comment by jotak** [comment]
@lucasponce I need to set up the route. Doing it now.

**Comment by jotak** [comment]
@lucasponce I've updated the PR but it's still stalled because of dependency on SWS-62 (rest endpoints) and missing tests ('til we all agree on the framework to use)

Since I don't think any other task is depending on this one, I think there's no problem in keeping it unmerged til it's all done.

**Comment by pilhuhn** [comment]
@jotak Test framework has been determined to be Testify

File: prometheus/client.go [hunk 8]
```diff
@@ -31,23 +31,27 @@ func NewClient() (*Client, error) {
...
+func (in *Client) inject(api v1.API) {
+	in.api = api
+}
+
 // GetSourceServices returns a map of source services for a given service identified by its namespace and service name.
 // Returned map has a destination version as a key and a "<origin service>/<origin version>" pair as value.
 // Destination service is not included in the map as it is passed as argument.
 // It returns an error on any problem.
-func (in *Client) GetSourceServices(namespace string, servicename string) (map[string]string, error) {
+func (in *Client) GetSourceServices(namespace string, servicename string) (map[string][]string, error) {
```
Comment (jotak) [inline prometheus/client.go]: @lucasponce I did some changes in the prometheus client, can you please review:

- GetSourceServices now returns a "multimap", since there can be more than 1 source service per version.
- The client now creates the API object from the start and keeps it in memory (thus it doesn't need to store the internal `client` field). This is the way I found to be able to inject and mock API from tests. Maybe there are other ways to do, I don't know.

**Comment by jotak** [comment]
TODO list is complete, so for my side it's ready to merge.

File: prometheus/client.go [hunk 9]
```diff
@@ -33,40 +31,72 @@ func NewClient() (*Client, error) {
...
+
 // GetSourceServices returns a map of source services for a given service identified by its namespace and service name.
 // Returned map has a destination version as a key and a "<origin service>/<origin version>" pair as value.
 // Destination service is not included in the map as it is passed as argument.
 // It returns an error on any problem.
-func (in *Client) GetSourceServices(namespace string, servicename string) (map[string]string, error) {
-	query := fmt.Sprintf(istioRequestQuery, servicename, namespace)
-	api := v1.NewAPI(in.p8s)
-	result, err := api.Query(context.Background(), query, time.Now())
+func (in *Client) GetSourceServices(namespace string, servicename string) (map[string][]string, error) {
```
Comment (lucasponce) [inline prometheus/client.go]: Please, update the comments describing the signature as return type has changed from map[string]string to map[string][]string so here should be changed:
https://github.com/swift-sunshine/swscore/pull/30/files#diff-970c77c7471b416af8e59de92b0accc0R44

File: prometheus/client.go [hunk 10]
```diff
@@ -33,40 +31,72 @@ func NewClient() (*Client, error) {
...
 			index := fmt.Sprintf("%s", metric["destination_version"])
 			sourceService := string(metric["source_service"])
-			// .svc sufix is a pure Istio label, I guess we can skip it at the moment for clarity
-			if i := strings.Index(sourceService, ".svc"); i > 0 {
+			// sourceService is in the form "service.namespace.istio_identity_domain". We want to keep only "service.namespace".
+			if i := strings.Index(sourceService, "."+config.Get().IstioIdentityDomain); i > 0 {
 				sourceService = sourceService[:i]
 			}
-			routes[index] = fmt.Sprintf("%s/%s", sourceService, metric["source_version"])
+			source := fmt.Sprintf("%s/%s", sourceService, metric["source_version"])
```
Comment (lucasponce) [inline prometheus/client.go]: +1
These change allows to cover cases where we might have
```
ProductServices
Destination    Sources
v1                   Review/v1, Review/v2
v2                   Review/v3, Review/v4
```
So before we cover only
```
Destination    Sources
v1                   Review/v1
v2                   Review/v3
```
Which misses potential versions for a single destination.
Good change.

File: prometheus/client_test.go [hunk 11]
```diff
@@ -1,51 +1,231 @@
...
+		&model.Sample{
+			Metric: rqCustV2,
+			Value:  1},
+		&model.Sample{
+			Metric: rqCustV2ToV2,
+			Value:  2}}
+
+	client, api, err := setupMocked()
+	if err != nil {
+		t.Error(err)
```
Comment (pilhuhn) [inline prometheus/client_test.go]: Would it be better to use an assert statement here as well? `assert.NotNil(t, err) ` ?

File: prometheus/client.go [hunk 12]
```diff
@@ -33,40 +31,72 @@ func NewClient() (*Client, error) {
 	if err != nil {
 		return nil, err
 	}
-	client.p8s = p8s
+	client.api = v1.NewAPI(p8s)
```
Comment (lucasponce) [inline prometheus/client.go]: I understand the change, and perhaps with the prometheus client it could make sense as it stateless, but maintain the client could help to close() it or performs another clean operation.
Perhaps for this is ok.
So, no a blocker, just leaving a comment here about the change just in case in the future we may need to review it.

File: prometheus/client_test.go (hunk 11 above)
Comment (jotak) [reply prometheus/client_test.go]: Hmm... I guess you meant `assert.Nil(t, err)` instead of the "if" block?
I'll do it.

File: prometheus/client.go (hunk 12 above)
Comment (jotak) [reply prometheus/client.go]: I can put it back, np

**Comment by lucasponce** [comment]
Vendoring should be splitted in a separate PR as we commented when introduced the initial deps.

File: prometheus/client_test.go (hunk 11 above)
Comment (pilhuhn) [reply prometheus/client_test.go]: Actually I meant `NotNil` as you want to assert that err is not `nil`.
See also https://godoc.org/github.com/stretchr/testify/assert#NotNil

**Comment by jotak** [comment]
@lucasponce this is actually what I did, but `make dep-update` looks at all the imports in source and can add extra packages (transitive deps) that are not explicitly declared in glide.yaml.

That's why I came up with new packages despite there's already merged the PR that brings testify.
At least I put it in a separate commit.

File: prometheus/client_test.go (hunk 11 above)
Comment (jotak) [reply prometheus/client_test.go]: No, I don't want to assert it is not nil: I want to report it in case it is not nil, but the normal behaviour would be err == nil.

Comment (pilhuhn) [reply prometheus/client_test.go]: Yes you are right. I had a brain fart

Comment (jotak) [reply prometheus/client_test.go]: :-)

**Comment by jotak** [comment]
@lucasponce comments fixed

**Comment by lucasponce** [comment]
It looks ok.
I still think that third party dependencies should go in a different PR, as we defined, yes, I hit same scenario when I imported the clients, but then logic could be more clear and we can separate things.
Also, perhaps at some point we would like to squash all minor PRs into one labeled with the JIRA, so for that, splitting PRs may sense.

But I won't block this for that. if this is ok for you I will merge.

**Comment by jotak** [comment]
Ok, I didn't understand the process correctly. Basically what I did was:
- based on `master` modify glide.yaml, run `make dep-update`, commit and push
- then work on the feature.

Whereas the good procedure would actually be:
- work on the feature, modify glide.yaml and run `make dep-update`
- commit glide+vendor, checkout master and cherry-pick the commit, push
- rebase the feature branch, push


PS: ok for me if you merge. I'll do it correctly next time :)

**Comment by lucasponce** [comment]
When vendoring updated I guess we can:
- Just add the vendoring + glide files and prepare a commit in a separate PR with proper JIRA.
- Rebase and continue working in the feature.
- If the feature has minor commits to address minor changes, I think is better to squash/amend and put them in a single commit with the proper JIRA.

I think at the end we will create a clean history and we separate better the concepts, and we can analyze if a PR with 8 commits is due the JIRA is big, or it is correct and decide the task granularity in general.

That's my 2 cents.

Anyway. Merging.

**Comment by lucasponce** [comment]
Sorry, github reports it needs a rebase.

**Comment by jotak** [comment]
rebased
